buidl/taproot.py

```diff
@@ -265,15 +271,40 @@ def __init__(self, points, locktime=None, sequence=None):
 
     def get_tweak_point(self, tweak):
         if self.point.parity:
-            return -1 * self.point + tweak
+            tweak_point = -1 * self.point + tweak
         else:
-            return self.point + tweak
+            tweak_point = self.point + tweak
+        return tweak_point
+
+    def generate_nonces(self):
+        k_1, k_2 = randbelow(N), randbelow(N)
+        r_1, r_2 = k_1 * G, k_2 * G
+        return (k_1, k_2), (r_1, r_2)
+
+    def nonce_sums(self, nonce_point_pairs):
+        sum_1 = S256Point.combine([n[0] for n in nonce_point_pairs])
+        sum_2 = S256Point.combine([n[1] for n in nonce_point_pairs])
+        return sum_1, sum_2
+
+    def compute_coefficient(self, nonce_sums, sig_hash):
+        bytes_to_hash = (
+            nonce_sums[0].sec() + nonce_sums[1].sec() + self.point.bip340() + sig_hash
+        )
+        return big_endian_to_int(hash_musignonce(bytes_to_hash))
+
+    def compute_k(self, nonce_secrets, nonce_sums, sig_hash):
+        h = self.compute_coefficient(nonce_sums, sig_hash)
+        return (nonce_secrets[0] + h * nonce_secrets[1]) % N
+
+    def compute_r(self, nonce_sums, sig_hash):
+        h = self.compute_coefficient(nonce_sums, sig_hash)
+        return S256Point.combine([nonce_sums[0], h * nonce_sums[1]])
```

Comment:
The beauty of such well-structure code is how incredibly easy to read the above methods are. Nicely done!